Lack of autopatch transmit audio for half duplex node

Greetings, 

I am having difficulty getting any audio from the autopatch to transmit on my simplex (half-duplex) node. 

The app_rpt code is a very recent pull (ie, yesterday), freshly compiled with some additional debug text. 

The node uses usbradio, and duplex is set to "1". I had tried setting the vox parameters in rpt.conf, but ultimately decided to not set them and use their default values for now. 

Here is the debug from a call. I spoke into the phone during the call and the transmitter did not key. I am able to hear the radio's receive audio on the phone when the squelch/COS is opened, as expected. So receive audio is working, PTT and transmit audio seem to not be working. 

Thanks for any help you can provide, 

--E
de W6EL

```
simplex900*CLI> rpt fun 617081 *616265551212
[2024-08-23 11:07:46.026] DEBUG[41054] app_rpt/rpt_telemetry.c: cancel_pfxfone!!
[2024-08-23 11:07:46.026] DEBUG[41054] app_rpt.c: digits=6  source=0
[2024-08-23 11:07:46.136] DEBUG[41054] app_rpt/rpt_telemetry.c: cancel_pfxfone!!
[2024-08-23 11:07:46.136] DEBUG[41054] app_rpt.c: digits=61  source=0
[2024-08-23 11:07:46.136] DEBUG[41054] app_rpt.c: @@@@ action: autopatchup, param = noct=1,farenddisconnect=1,dialtime=20000,quiet=1,context=voipmscontext
[2024-08-23 11:07:46.136] DEBUG[41054] app_rpt.c: @@@@ table index i = 1
[2024-08-23 11:07:46.136] DEBUG[41054] app_rpt/rpt_functions.c: @@@@ Autopatch up
[2024-08-23 11:07:46.136] NOTICE[41054] app_rpt/rpt_functions.c: Processing autopatch parameter list...
[2024-08-23 11:07:46.136] NOTICE[41054] app_rpt/rpt_functions.c: Setting patchquiet to value: 1
[2024-08-23 11:07:46.137] DEBUG[41054] app_rpt.c: rv=3
[2024-08-23 11:07:46.137] NOTICE[41054] app_rpt/rpt_vox.c: Running voxinit_rpt, enable value is: 1
[2024-08-23 11:07:46.137] NOTICE[41091] app_rpt.c: Requested channel DAHDI/pseudo-1844845188
[2024-08-23 11:07:46.137] DEBUG[41091] app_rpt/rpt_call.c: No CDR present on DAHDI/pseudo-1844845188
[2024-08-23 11:07:46.137] DEBUG[41091] app_rpt/rpt_bridging.c: Channel DAHDI/pseudo-1844845188 joining conference 1022
[2024-08-23 11:07:46.138] NOTICE[41091] app_rpt.c: Requested channel DAHDI/pseudo-1667367254
[2024-08-23 11:07:46.138] DEBUG[41091] app_rpt/rpt_call.c: No CDR present on DAHDI/pseudo-1667367254
[2024-08-23 11:07:46.138] NOTICE[41091] app_rpt.c: tzone: 0, patchquiet: 1, patchexten[0]: 0
[2024-08-23 11:07:46.138] DEBUG[41091] app_rpt/rpt_bridging.c: Channel DAHDI/pseudo-1667367254 joining conference 1022
[2024-08-23 11:07:46.138] NOTICE[41091] app_rpt.c: Added speaker
[2024-08-23 11:07:46.138] NOTICE[41091] app_rpt.c: did set_tone_zone
[2024-08-23 11:07:46.138] NOTICE[41091] app_rpt.c: About to do play_dialtone
[2024-08-23 11:07:46.138] NOTICE[41091] app_rpt.c: Completed play_dialtone stanzas
[2024-08-23 11:07:48.800] VERBOSE[41092][C-0000000b] pbx.c: Executing [6265551212@voipmscontext:1] Set("DAHDI/pseudo-1844845188", "CALLERID(all)="W6EL" <7475551234>") in new stack
[2024-08-23 11:07:48.800] VERBOSE[41092][C-0000000b] pbx.c: Executing [6265551212@voipmscontext:2] Dial("DAHDI/pseudo-1844845188", "IAX2/voipms/+16265551212") in new stack
[2024-08-23 11:07:48.801] VERBOSE[41092][C-0000000b] app_dial.c: Called IAX2/voipms/+16265551212
[2024-08-23 11:07:48.809] NOTICE[41091] app_rpt/rpt_bridging.c: Starting rpt_call_bridge_setup...
[2024-08-23 11:07:48.809] NOTICE[41091] app_rpt/rpt_bridging.c: Running rpt_conf_add_speaker
[2024-08-23 11:07:48.809] DEBUG[41091] app_rpt/rpt_bridging.c: Channel DAHDI/pseudo-391496481 joining conference 1022
[2024-08-23 11:07:48.809] DEBUG[41091] app_rpt/rpt_bridging.c: Channel DAHDI/pseudo-1195219378 joining conference 1022
[2024-08-23 11:07:48.810] NOTICE[41091] app_rpt/rpt_bridging.c: Finished rpt_call_bridge_setup, returning 0.
[2024-08-23 11:07:48.944] VERBOSE[41043][C-0000000b] chan_iax2.c: Call accepted by 208.100.60.36:4569 (format ulaw)
[2024-08-23 11:07:48.944] VERBOSE[41043][C-0000000b] chan_iax2.c: Format for call is (ulaw)
[2024-08-23 11:07:50.920] VERBOSE[41092][C-0000000b] app_dial.c: IAX2/voipms-10234 is making progress passing it to DAHDI/pseudo-1844845188
[2024-08-23 11:07:51.136] VERBOSE[41092][C-0000000b] app_dial.c: IAX2/voipms-10234 is making progress passing it to DAHDI/pseudo-1844845188
[2024-08-23 11:07:55.564] VERBOSE[41092][C-0000000b] app_dial.c: IAX2/voipms-10234 answered DAHDI/pseudo-1844845188
[2024-08-23 11:07:55.565] VERBOSE[41094][C-0000000b] bridge_channel.c: Channel IAX2/voipms-10234 joined 'simple_bridge' basic-bridge <ce232dd3-f4a9-494e-8ddc-fd910295bb86>
[2024-08-23 11:07:55.565] VERBOSE[41092][C-0000000b] bridge_channel.c: Channel DAHDI/pseudo-1844845188 joined 'simple_bridge' basic-bridge <ce232dd3-f4a9-494e-8ddc-fd910295bb86>
[2024-08-23 11:08:00.076] DEBUG[41054] app_rpt.c: Time now: 11:08 23 08 05

; Here is where I am speaking into the phone for about 15 seconds. 
; And then I hang up from the phone.

[2024-08-23 11:08:15.454] VERBOSE[41094][C-0000000b] bridge_channel.c: Channel IAX2/voipms-10234 left 'simple_bridge' basic-bridge <ce232dd3-f4a9-494e-8ddc-fd910295bb86>
[2024-08-23 11:08:15.455] VERBOSE[41092][C-0000000b] bridge_channel.c: Channel DAHDI/pseudo-1844845188 left 'simple_bridge' basic-bridge <ce232dd3-f4a9-494e-8ddc-fd910295bb86>
[2024-08-23 11:08:15.455] VERBOSE[41094][C-0000000b] chan_iax2.c: Hungup 'IAX2/voipms-10234'
[2024-08-23 11:08:15.455] VERBOSE[41092][C-0000000b] pbx.c: Spawn extension (voipmscontext, 6265551212, 2) exited non-zero on 'DAHDI/pseudo-1844845188'
[2024-08-23 11:08:15.455] VERBOSE[41092][C-0000000b] chan_dahdi.c: Hungup 'DAHDI/pseudo-1844845188'
[2024-08-23 11:08:15.460] DEBUG[41091] app_rpt.c: callmode=3, patchfarenddisconnect=1, duplex=1
[2024-08-23 11:08:15.473] NOTICE[41054] app_rpt/rpt_vox.c: Running voxinit_rpt, enable value is: 0
[2024-08-23 11:08:15.480] DEBUG[41091] app_rpt.c: exit channel loop
[2024-08-23 11:08:15.480] VERBOSE[41091] chan_dahdi.c: Hungup 'DAHDI/pseudo-1667367254'
[2024-08-23 11:08:15.480] DEBUG[41091] app_rpt/rpt_rig.c: remoterig=, nowchan=00, waschan=00
[2024-08-23 11:08:15.481] DEBUG[41091] app_rpt/rpt_bridging.c: Channel DAHDI/pseudo-391496481 joining conference 1022

```

I ask because asterisk isn't building, but it's an error which I remember was tied with a mismatch in dahdi versions. 

```
chan_dahdi.c: In function ‘dahdievent_to_analogevent’:
chan_dahdi.c:2669:14: error: ‘DAHDI_EVENT_PULSE’ undeclared (first use in this function); did you mean ‘DAHDI_EVENT_NONE’?
 2669 |         case DAHDI_EVENT_PULSE:
      |              ^~~~~~~~~~~~~~~~~
      |              DAHDI_EVENT_NONE
chan_dahdi.c:2669:14: note: each undeclared identifier is reported only once for each function it appears in
chan_dahdi.c:2672:14: error: ‘DAHDI_EVENT_PULSE_BREAK’ undeclared (first use in this function); did you mean ‘ANALOG_EVENT_PULSE_BREAK’?
 2672 |         case DAHDI_EVENT_PULSE_BREAK:
      |              ^~~~~~~~~~~~~~~~~~~~~~~
      |              ANALOG_EVENT_PULSE_BREAK
chan_dahdi.c: In function ‘dahdi_handle_event’:
chan_dahdi.c:7907:14: error: ‘DAHDI_EVENT_PULSE’ undeclared (first use in this function); did you mean ‘DAHDI_EVENT_NONE’?
 7907 |         case DAHDI_EVENT_PULSE:
      |              ^~~~~~~~~~~~~~~~~
      |              DAHDI_EVENT_NONE
chan_dahdi.c:7908:14: error: ‘DAHDI_EVENT_PULSE_BREAK’ undeclared (first use in this function); did you mean ‘ANALOG_EVENT_PULSE_BREAK’?
 7908 |         case DAHDI_EVENT_PULSE_BREAK:
      |              ^~~~~~~~~~~~~~~~~~~~~~~
      |              ANALOG_EVENT_PULSE_BREAK
```

Not your error at all, just some configuration stuff to contend with of course.

--- a/apps/app_rpt/rpt_bridging.c
+++ b/apps/app_rpt/rpt_bridging.c
@@ -417,23 +417,20 @@ static int *dahdi_confno(struct rpt *myrpt, enum rpt_conf_type type)
 }
 
 /*!
- * \brief Get the conference number of a DAHDI channel
- * This actually gets the internal DAHDI channel number.
- * The channel number is used by the DAHDI monitor functions
- * to create a monitor only conference.
+ * \brief Get the channel number of a DAHDI channel
  * \param chan DAHDI channel
  * \retval -1 on failure, conference number on success
  */
-static int dahdi_conf_fd_confno(struct ast_channel *chan)
+static int dahdi_conf_get_channo(struct ast_channel *chan)
 {
 	struct dahdi_confinfo ci = {0};
 
-	if (ioctl(ast_channel_fd(chan, 0), DAHDI_CHANNO, &ci.confno)) {
+	if (ioctl(ast_channel_fd(chan, 0), DAHDI_CHANNO, &ci.chan)) {
 		ast_log(LOG_WARNING, "DAHDI_CHANNO failed: %s\n", strerror(errno));
 		return -1;
 	}
 
-	return ci.confno;
+	return ci.chan;
 }
 
 int __rpt_conf_create(struct ast_channel *chan, struct rpt *myrpt, enum rpt_conf_type type, enum rpt_conf_flags flags, const char *file, int line)
@@ -488,7 +485,7 @@ int rpt_call_bridge_setup(struct rpt *myrpt, struct ast_channel *mychannel, stru
 	}
 
 	/* get its conference number */
-	res = dahdi_conf_fd_confno(mychannel);
+	res = dahdi_conf_get_channo(mychannel);
 	if (res < 0) {
 		ast_log(LOG_WARNING, "Unable to get autopatch channel number\n");
 		ast_hangup(mychannel);
@@ -498,10 +495,13 @@ int rpt_call_bridge_setup(struct rpt *myrpt, struct ast_channel *mychannel, stru
 	/* put vox channel monitoring on the channel  
 	 *
 	 * This uses the internal DAHDI channel number to create the 
-	 * monitor conference.  It requires a patched version of DAHDI
-	 * that contains changes for this to work properly.  
-	 * Without the patch, this code will hang here when trying to 
-	 * join the conference.
+	 * monitor conference.  This code will hang here when trying to 
+	 * join the conference when the underlying version of DAHDI in use
+	 * is missing a patch that allows the DAHDI_CONF_MONITOR option
+	 * to monitor a pseudo channel.  This patch prevents the hardware
+	 * pre-echo routines from acting on a pseudo channel.  It also
+	 * prevents the DAHDI check conference routine from acting
+	 * on a channel number being used as a conference.
 	 */
 	if (dahdi_conf_add(myrpt->voxchannel, res, DAHDI_CONF_MONITOR)) {
 		ast_hangup(mychannel);
@@ -515,7 +515,7 @@ int rpt_mon_setup(struct rpt *myrpt)
 	int res;
 
 	if (!IS_PSEUDO(myrpt->txchannel) && myrpt->dahditxchannel == myrpt->txchannel) {
-		int confno = dahdi_conf_fd_confno(myrpt->txchannel); /* get tx channel's port number */
+		int confno = dahdi_conf_get_channo(myrpt->txchannel); /* get tx channel's port number */
 		if (confno < 0) {
 			return -1;
 		}

--- a/apps/app_rpt/rpt_bridging.c
+++ b/apps/app_rpt/rpt_bridging.c
@@ -418,6 +418,9 @@ static int *dahdi_confno(struct rpt *myrpt, enum rpt_conf_type type)
 
 /*!
  * \brief Get the conference number of a DAHDI channel
+ * This actually gets the internal DAHDI channel number.
+ * The channel number is used by the DAHDI monitor functions
+ * to create a monitor only conference.
  * \param chan DAHDI channel
  * \retval -1 on failure, conference number on success
  */
@@ -428,13 +431,7 @@ static int dahdi_conf_fd_confno(struct ast_channel *chan)
 	ci.confno = 0;
 	ci.confmode = 0;
 
-	/* This routine previously called DAHDI_CHANNO.  After testing, we 
-	 * determined that DAHDI_CHANNO is actually returning the channel number, 
-	 * instead of the conference number.  In some cases the channel number
-	 * was same as the conference number.  Since it was not accurate in
-	 * all cases, it was switched to DAHDI_GETCONF.
-	 */
-	if (ioctl(ast_channel_fd(chan, 0), DAHDI_GETCONF, &ci)) {
+	if (ioctl(ast_channel_fd(chan, 0), DAHDI_CHANNO, &ci)) {
 		ast_log(LOG_WARNING, "DAHDI_GETCONF failed: %s\n", strerror(errno));
 		return -1;
 	}
@@ -503,12 +500,13 @@ int rpt_call_bridge_setup(struct rpt *myrpt, struct ast_channel *mychannel, stru
 
 	/* put vox channel monitoring on the channel  
 	 *
-	 * The conference flags were originally DAHDI_CONF_MONITOR.  This 
-	 * setting caused dahdi_conf_add to return EINVAL.  The  flags
-	 * RPT_CONF_CONF | RPT_CONF_LISTENER permits the vox channel to 
-	 * join the conference.
+	 * This uses the internal DAHDI channel number to create the 
+	 * monitor conference.  It requires the ASL version of DAHDI
+	 * that contains changes for this to work properly.  
+	 * If the standard Asterisk DAHDI driver is used, this
+	 * code will hang here when trying to join the conference.
 	 */
-	if (dahdi_conf_add(myrpt->voxchannel, res, RPT_CONF_CONF | RPT_CONF_LISTENER)) {
+	if (dahdi_conf_add(myrpt->voxchannel, res, DAHDI_CONF_MONITOR)) {
 		ast_hangup(mychannel);
 		return -1;
 	}

--- a/apps/app_rpt/rpt_bridging.c
+++ b/apps/app_rpt/rpt_bridging.c
@@ -431,7 +431,7 @@ static int dahdi_conf_fd_confno(struct ast_channel *chan)
 	ci.confno = 0;
 	ci.confmode = 0;
 
-	if (ioctl(ast_channel_fd(chan, 0), DAHDI_CHANNO, &ci)) {
+	if (ioctl(ast_channel_fd(chan, 0), DAHDI_CHANNO, &ci.confno)) {
 		ast_log(LOG_WARNING, "DAHDI_GETCONF failed: %s\n", strerror(errno));
 		return -1;
 	}

--- a/apps/app_rpt/rpt_bridging.c
+++ b/apps/app_rpt/rpt_bridging.c
@@ -426,13 +426,10 @@ static int *dahdi_confno(struct rpt *myrpt, enum rpt_conf_type type)
  */
 static int dahdi_conf_fd_confno(struct ast_channel *chan)
 {
-	struct dahdi_confinfo ci;
-	ci.chan = 0;
-	ci.confno = 0;
-	ci.confmode = 0;
+	struct dahdi_confinfo ci = {0};
 
 	if (ioctl(ast_channel_fd(chan, 0), DAHDI_CHANNO, &ci.confno)) {
-		ast_log(LOG_WARNING, "DAHDI_GETCONF failed: %s\n", strerror(errno));
+		ast_log(LOG_WARNING, "DAHDI_CHANNO failed: %s\n", strerror(errno));
 		return -1;
 	}
 
@@ -501,10 +498,10 @@ int rpt_call_bridge_setup(struct rpt *myrpt, struct ast_channel *mychannel, stru
 	/* put vox channel monitoring on the channel  
 	 *
 	 * This uses the internal DAHDI channel number to create the 
-	 * monitor conference.  It requires the ASL version of DAHDI
+	 * monitor conference.  It requires a patched version of DAHDI
 	 * that contains changes for this to work properly.  
-	 * If the standard Asterisk DAHDI driver is used, this
-	 * code will hang here when trying to join the conference.
+	 * Without the patch, this code will hang here when trying to 
+	 * join the conference.
 	 */
 	if (dahdi_conf_add(myrpt->voxchannel, res, DAHDI_CONF_MONITOR)) {
 		ast_hangup(mychannel);

